### PR TITLE
Couch to SQL: mostly changes for case diffs

### PR DESCRIPTION
**Pull request by millerdev**
Review by commit 🐡

### CODE REVIEW

File: corehq/apps/couch_sql_migration/rebuildcase.py [hunk 1]
```diff
@@ -18,17 +18,19 @@ def is_action_order_equal(sql_case, couch_json):
     """
     def dedup(items):
         return list(dict.fromkeys(items))
+    if was_rebuilt(sql_case):
+        return False
     sql_ids = dedup(t.form_id for t in sql_case.transactions if t.form_id)
     couch_ids = dedup(a["xform_id"] for a in couch_json["actions"] if a["xform_id"])
-    return sql_ids == couch_ids
+    return len(sql_ids) == len(couch_ids) and sql_ids != couch_ids
```
Comment (dannyroberts) [inline corehq/apps/couch_sql_migration/rebuildcase.py]: I'm a little confused how it went from `sql_ids == couch_ids` to `<something> and sql_ids != couch_ids` without nearly reversing the meaning, maybe you could just fill in what I'm not seeing?

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 2]
```diff
@@ -946,7 +946,18 @@ def data_function(**view_kwargs):
         item_getter=None,
         event_handler=MigrationPaginationEventHandler(domain, stopper)
     )
-    return (row[row_key] for row in rows)
+    log.info("iteration state: %r", rows.state)
+    row = None
+    try:
+        for row in rows:
+            assert row['key'][0] == domain, row
```
Comment (dannyroberts) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: Is this the thing we talked about that will make the case diff process fail hard if it ever is about to process a case from the wrong domain?

File: corehq/apps/couch_sql_migration/rebuildcase.py (hunk 1 above)
Comment (dannyroberts) [reply corehq/apps/couch_sql_migration/rebuildcase.py]: Is this where you implement the item
- do not do SQL rebuild if transactions differ between couch and SQL
we talked about earlier?

I think this line is supposed to check that it's the same ids but just in a different order, which isn't strictly what it does. If I'm getting the intent right, maybe it's worth changing this to

```
    return set(sql_ids) == set(couch_ids) and sql_ids != couch_ids
```
?

**Review by dannyroberts — APPROVED** [review]

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 2 above)
Comment (millerdev) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Correct

File: corehq/apps/couch_sql_migration/rebuildcase.py (hunk 1 above)
Comment (millerdev) [reply corehq/apps/couch_sql_migration/rebuildcase.py]: Good catch. I'll change that.

File: corehq/apps/couch_sql_migration/casediff.py [hunk 3]
```diff
@@ -732,12 +738,16 @@ def diff_case(sql_case, couch_case, dd_count):
     def diff(couch_json, sql_json):
         diffs = json_diff(couch_json, sql_json, track_list_indices=False)
         return filter_case_diffs(couch_json, sql_json, diffs, _diff_state)
+    case_id = couch_case['_id']
     sql_json = sql_case.to_json()
     dd_count("commcare.couchsqlmigration.case.diffed")
+    diffs = check_domains(case_id, couch_case, sql_json)
+    if diffs:
+        dd_count("commcare.couchsqlmigration.case.has_diff")
+        return couch_case, diffs
```
Comment (millerdev) [inline corehq/apps/couch_sql_migration/casediff.py]: Check domain just before the case will be diffed and possibly rebuilt. It turned out to require more code than I had hoped, although it allowed a few other nice cleanups now that the domain is on the statedb. This makes it much harder to imagine a scenario where the diff code would accidentally modify documents in the wrong domain.

**Review by dannyroberts — APPROVED** [review]

**Pull request merged**